File: kata-03-claude-sdk-tools-ts/starter.ts
/**
 * Kata 03: Claude Agent SDK with Custom Tools (TypeScript) — Starter
 *
 * Fill in the TODOs. Compare with solution.ts when stuck.
 *
 *   1. Define tools with tool(name, description, zodShape, async handler).
 *   2. Bundle them in createSdkMcpServer({ name, version, tools }).
 *   3. Pass mcpServers + allowedTools (`mcp__<server>__<tool>`) in Options.
 */

import "dotenv/config";
import { z } from "zod";
import {
  query,
  tool,
  createSdkMcpServer,
  type Options,
  type SdkMcpToolDefinition,
} from "@anthropic-ai/claude-agent-sdk";

const textResult = (text: string) => ({ content: [{ type: "text" as const, text }] });

// ---------------------------------------------------------------------------
// TODO 1: get_current_time
// ---------------------------------------------------------------------------
// Zod shape: { tz_name: z.string().default("UTC") }
// Return current ISO timestamp (or formatted UTC).
const getCurrentTime = tool(
  "get_current_time",
  "TODO description",
  { tz_name: z.string().default("UTC") },
  async ({ tz_name }) => {
    throw new Error("TODO 1: implement get_current_time");
  }
);

// ---------------------------------------------------------------------------
// TODO 2: calculate
// ---------------------------------------------------------------------------
// Accept { expression: z.string() }, validate it contains only safe chars
// ([0-9+\-*/().\s]), evaluate it with `new Function(...)` and return the result.
const calculate = tool(
  "calculate",
  "TODO description",
  { expression: z.string() },
  async ({ expression }) => {
    throw new Error("TODO 2: implement calculate");
  }
);

// ---------------------------------------------------------------------------
// TODO 3: wire the MCP server and options
// ---------------------------------------------------------------------------
// - server = createSdkMcpServer({ name: "workshop", version: "1.0.0", tools: [...] })
// - In Options:
//     mcpServers: { workshop: server }
//     allowedTools: ["mcp__workshop__get_current_time", "mcp__workshop__calculate"]
function buildOptions(): Options {
  throw new Error("TODO 3: build options");
}

// ---------------------------------------------------------------------------
// TODO 4: run a query and print text + tool calls
// ---------------------------------------------------------------------------
async function run(prompt: string) {
  const options = buildOptions();
  for await (const msg of query({ prompt, options })) {
    if (msg.type === "assistant") {
      for (const block of msg.message.content) {
        if (block.type === "text" && block.text.trim()) {
          console.log(`Agent: ${block.text}`);
        } else if (block.type === "tool_use") {
          console.log(`  → tool: ${block.name}(${JSON.stringify(block.input)})`);
        }
      }
    }
  }
}

async function main() {
  await run("What time is it, and what is 12 * 9 + 3?");
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: kata-03-claude-sdk-tools/starter.py
"""
Kata 03: Claude Agent SDK with Custom Tools — Starter

Fill in the TODOs to build an agent that calls your custom tools.

The shape:
    1. Decorate async functions with @tool(name, description, schema).
    2. Wrap them in an MCP server with create_sdk_mcp_server(...).
    3. Reference the server in ClaudeAgentOptions(mcp_servers={...}).
    4. Allowlist tools with allowed_tools=["mcp__<server>__<tool>", ...].
"""

import asyncio
from datetime import datetime, timezone
from typing import Any

from dotenv import load_dotenv

from claude_agent_sdk import (
    AssistantMessage,
    ClaudeAgentOptions,
    TextBlock,
    ToolUseBlock,
    create_sdk_mcp_server,
    query,
    tool,
)

load_dotenv()


def text_result(s: str) -> dict[str, Any]:
    return {"content": [{"type": "text", "text": s}]}


# -----------------------------------------------------------------------------
# TODO 1: define a `get_current_time` tool
# -----------------------------------------------------------------------------
# Use the @tool(name, description, schema) decorator. Schema is a dict
# mapping arg name to type (e.g. {"tz_name": str}). The async function
# receives `args: dict` and must return text_result(...).
@tool("get_current_time", "TODO description", {"tz_name": str})
async def get_current_time(args: dict[str, Any]) -> dict[str, Any]:
    raise NotImplementedError("TODO 1: implement get_current_time")


# -----------------------------------------------------------------------------
# TODO 2: define a `calculate` tool
# -----------------------------------------------------------------------------
# Accept an expression string and return its evaluated value. Use a
# restricted eval (no __builtins__) — see solution.py for the safe-eval
# pattern.
@tool("calculate", "TODO description", {"expression": str})
async def calculate(args: dict[str, Any]) -> dict[str, Any]:
    raise NotImplementedError("TODO 2: implement calculate")


# -----------------------------------------------------------------------------
# TODO 3: wire up the MCP server and options
# -----------------------------------------------------------------------------
# - Build server = create_sdk_mcp_server("workshop", "1.0.0", [tools...])
# - In options, set:
#     mcp_servers={"workshop": server}
#     allowed_tools=["mcp__workshop__get_current_time", "mcp__workshop__calculate"]
def build_options() -> ClaudeAgentOptions:
    raise NotImplementedError("TODO 3: build options with MCP server + allowed_tools")


# -----------------------------------------------------------------------------
# TODO 4: run a query and print both text and tool calls
# -----------------------------------------------------------------------------
async def run(prompt: str) -> None:
    options = build_options()
    async for msg in query(prompt=prompt, options=options):
        if isinstance(msg, AssistantMessage):
            for block in msg.content:
                if isinstance(block, TextBlock) and block.text.strip():
                    print(f"Agent: {block.text}")
                elif isinstance(block, ToolUseBlock):
                    print(f"  → tool: {block.name}({block.input})")


async def main():
    await run("What time is it, and what is 12 * 9 + 3?")


if __name__ == "__main__":
    asyncio.run(main())
